### PR TITLE
Feature/farm 481 firebase profile repo

### PR DESCRIPTION
**Pull request by LeeMatthewHiggins**
[FARM-481 - firebase profile repo]

#### 📲 What

Implements the profile repository for firebase.
App now persists login and profiles.

#### 🤔 Why
		
Why it's needed, To persist profiles on the server

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+          }
+          return null;
+        });
+      }
+    });
+  }
+
+  @override
+  Future<ProfileEntity> getSingle(String uri) {
+    // TODO: implement getSingle
```
Comment (LeeMatthewHiggins) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart:105]: these are not used atm, I will come back and remove if we don't need them

File: lib/model/bloc/chatFlow/NewProfileFlow.dart [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  FlowCoordinatorStatus _status = FlowCoordinatorStatus.Idle;
+
+  NewProfileFlowCoordinator(this._accountRepository, Function onStatusChanged): this._onStatusChanged = onStatusChanged;
+  
+  void run(BuildContext context, {Function onSuccess, Function onFail}){
+      _setStatus(FlowCoordinatorStatus.InProgress);
+       NavigationScope.presentModal(
+        context, ChatPage(viewModel: _chatPageViewModel(onSuccess: () {
+          _setStatus(FlowCoordinatorStatus.Complete);
+          return null;
```
Comment (alexanishchenko88) [inline lib/model/bloc/chatFlow/NewProfileFlow.dart]: Not sure about that, but is this return needed?

File: lib/model/bloc/profile/PersonName.dart [hunk 3]
```diff
@@ -44,7 +44,7 @@ class PersonName {
               final firstName = (names.length > 0) ? names.first : _Strings.empty;
               final secondName = (names.length > 1) ? names[1] : _Strings.empty;
               final firstInitial =  firstName.isNotEmpty ? firstName.substring(0, 1) : _Strings.empty;
-              final secondInitial = secondName.isNotEmpty ? (_Strings.divider + secondName.substring(0,1)) : _Strings.empty;
+              final secondInitial = secondName.isNotEmpty ? secondName.substring(0, 1) : _Strings.empty;
```
Comment (alexanishchenko88) [inline lib/model/bloc/profile/PersonName.dart:47]: No divider between first and second name?

File: lib/model/bloc/startup/StartupViewModelProvider.dart [hunk 4]
```diff
@@ -26,46 +22,57 @@ class _LocalisedStrings {
...
-          (account != null),
-          _landingViewModel(),
-        );
-        _controller.sink.add(_snapshot);
+        if (account != null) {
+          account.profileRepository.observeCurrent().listen((currentProfile) {
+            _updateState(currentProfile);
+          });
+          account.profileRepository.getCurrent().then((currentProfile) {
+            _updateState(currentProfile);
```
Comment (alexanishchenko88) [inline lib/model/bloc/startup/StartupViewModelProvider.dart:53]: What is the difference between observeCurrent and get current. The callback is doing the same for both.

File: lib/model/entities/ProfileEntity.dart [hunk 5]
```diff
@@ -6,5 +6,5 @@ class ProfileEntity {
     final ImageURLProvider avatar;
     final Map<String,String> lastPlotInfo;
 
-  ProfileEntity(this.id, this.name, this.avatar, this.lastPlotInfo);
+  ProfileEntity(this.id, this.name, this.avatar, this.lastPlotInfo,);
```
Comment (alexanishchenko88) [inline lib/model/entities/ProfileEntity.dart]: Even with come at the and somehow is not applying the format correctly 🤔

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+  @override
+  Future<ProfileEntity> add(ProfileEntity profile) {
+    return _user.then((user) {
+      final Map<String, dynamic> data =
+          _transformToFirebase.transform(from: profile);
+      return _profilesCollection(user).add(data).then((result) {
+        return result.get().then((document) {
+          final newProfile = _transformFromFirebase.transform(from: document);
+          return newProfile;
```
Comment (alexanishchenko88) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: Why creating this final variable if you just can:

return _transformFromFirebase.transform(from: document);

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        return true;
+      });
+    }, onError: (error) {
+      return false;
+    });
+  }
+
+  String _currentProfilePath(FirebaseUser user, String profileID) {
+    return _userPath(user) +
+        _Fields.seporator +
```
Comment (alexanishchenko88) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]:
```suggestion
        _Fields.separator +
```

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        _Fields.profiles +
+        _Fields.seporator +
```
Comment (alexanishchenko88) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]:
```suggestion
        _Fields.separator +
```

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        profileID;
+  }
+
+  String _userPath(FirebaseUser user) {
+    return _Fields.collectionName + _Fields.seporator + user.uid;
```
Comment (alexanishchenko88) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]:
```suggestion
    return _Fields.collectionName + _Fields.separator + user.uid;
```

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  CollectionReference _profilesCollection(FirebaseUser user) {
+    final path = _userPath(user) + _Fields.seporator + _Fields.profiles;
```
Comment (alexanishchenko88) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]:
```suggestion
    final path = _userPath(user) + _Fields.separator + _Fields.profiles;
```

File: lib/ui/LandingPage.dart [hunk 11]
```diff
@@ -213,8 +211,11 @@ class LandingPage extends StatelessWidget {
     BuildContext context,
     LandingPageViewModel viewModel,
   }) {
-    NavigationScope.presentModal(
-        context, ChatPage(viewModel: viewModel.chatViewModel));
+    viewModel.newAccountFlow.run(context, onSuccess: (){
+      print("YES");
```
Comment (alexanishchenko88) [inline lib/ui/LandingPage.dart]: Should be localised string?

**Review by alexanishchenko88 — COMMENTED** [review]
Reviewed!
I just left some minor comments.

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+import 'DocumentToProfileEntityTransformer.dart';
+import 'ProfileEntityToDocumentTransformer.dart';
+
+class _Fields {
+  static const collectionName = "fs_users";
+  static const ownerID = "owner";
+  static const profiles = "profiles";
+  static const currentProfile = "currentProfile";
+  static const seporator = "/";
```
Comment (davidfranquet) [inline lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]:
```suggestion
  static const separator = "/";
```

File: lib/ui/LandingPage.dart (hunk 11 above)
Comment (davidfranquet) [reply lib/ui/LandingPage.dart]: I don't think so

**Review by davidfranquet — COMMENTED** [review]
Just minors

File: lib/model/bloc/chatFlow/NewProfileFlow.dart (hunk 2 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/chatFlow/NewProfileFlow.dart]: its not needed. 👍

File: lib/model/bloc/profile/PersonName.dart (hunk 3 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/profile/PersonName.dart:47]: nope, this makes the string long and looks odd. I realised google and others don't have a space.

Comment (LeeMatthewHiggins) [reply lib/model/bloc/profile/PersonName.dart:47]: LH instead of L H

File: lib/model/bloc/startup/StartupViewModelProvider.dart (hunk 4 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/startup/StartupViewModelProvider.dart:53]: one observes for changes, the other activity triggers a refresh. I have thought about having a method which combines both.

File: lib/model/entities/ProfileEntity.dart (hunk 5 above)
Comment (LeeMatthewHiggins) [reply lib/model/entities/ProfileEntity.dart]: must have forgot to run the format

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: 👍

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 7 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: 👍

File: lib/ui/LandingPage.dart (hunk 11 above)
Comment (LeeMatthewHiggins) [reply lib/ui/LandingPage.dart]: that's debug code, I should remove. this callback if only for if you want to do something on the view layer, which we don't need in this case ATM.

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 12 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: done.

File: lib/ui/LandingPage.dart (hunk 11 above)
Comment (LeeMatthewHiggins) [reply lib/ui/LandingPage.dart]: sorted 👍

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 10 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: sorted 👍

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 9 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: sorted 👍

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 8 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: sorted 👍

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 7 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: sorted 👍

File: lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/model/repositories/profile/implementation/FirebaseProfileRepository.dart]: sorted 👍

File: lib/model/entities/ProfileEntity.dart (hunk 5 above)
Comment (LeeMatthewHiggins) [reply lib/model/entities/ProfileEntity.dart]: sorted 👍

File: lib/model/bloc/chatFlow/NewProfileFlow.dart (hunk 2 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/chatFlow/NewProfileFlow.dart]: sorted 👍

**Review by alexanishchenko88 — APPROVED** [review]

**Review by davidfranquet — APPROVED** [review]

**Pull request merged**